lib/spack/llnl/util/filesystem.py

```diff
@@ -2478,27 +2478,41 @@ def _link(self, path, dest_dir):
         This is because it is both meaningless from an rpath
         perspective, and will cause an error when Developer
         mode is not enabled"""
+
+        def report_already_linked():
+            already_linked = islink(dest_file)
+            tty.debug(
+                "Linking library %s to %s failed, " % (path, dest_file) + "already linked."
+                if already_linked
+                else "library with name %s already exists at location %s." % (file_name, dest_dir)
+            )
+
         file_name = os.path.basename(path)
         dest_file = os.path.join(dest_dir, file_name)
         if os.path.exists(dest_dir) and not dest_file == path:
```

Comment:
One option: can we modify this so it doesn't try if `dest_file` already exists (or, given the current logic, `lexists`, since it looks based on the specific exception handling that we allow not-a-link-but-a-file-that-exists errors to percolate up)?

Reply:
I will implement this in addition to the multiple new error types. 

For the sake of clarity we do not allow `not-a-link-but-a-file-that-exists` errors to percolate, they're caught by the same check that triggers the `SymlinkAlreadyExists` error since `lexists` return true for normal files as well, so this handling catches both scenarios since the `symlink` module does not differentiate.